In [2]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [4]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [5]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

predictors_norm = (predictors - predictors.mean()) / predictors.std()
target_norm = (target - target.mean()) / target.std()

X=predictors_norm
y=target_norm

n_cols = predictors_norm.shape[1] # number of predictors

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

## Build the Neural Network

In [6]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# build the model
model = regression_model()

## Fit the Model, Find the predictions, and Calculate the MSE

In [7]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=0)


y_pred=model.predict(X_test)
MSE=mean_squared_error(y_test, y_pred)

print(MSE)


0.20628705289978744


## Repeat the process for 50 times

In [8]:
MSE = 0
MSEs = []

for i in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    MSE = mean_squared_error(y_test, y_pred)
    print(MSE)
    MSEs.append(MSE)

59.51611075331527
61.240379341596636
53.56368776245942
49.70816598457399
65.13703788582008
44.748665616388685
57.8125977953775
42.1694758724032
44.045806414526346
49.14625773562669
43.73603068044583
46.995542121647105
41.22983954234689
44.1401264144703
38.594772793387314
43.21735852114742
35.289057141159766
40.82375649131288
41.918129019605075
35.6472264172967
45.126836983678565
41.509834685570866
34.60071485529688
42.28981220416108
38.586750191562544
38.63543776066299
39.36865787120252
44.64686471955762
44.63133147983098
40.59638917113744
38.41311165309203
33.83163826457487
33.88748309357244
36.41251187160466
40.41539428752437
32.96924169279539
38.282881782419814
36.500273316410194
36.433520413871385
31.840635661048985
42.57714342414739
33.39309779108059
35.266807235230644
42.58106691255029
37.49441331435595
40.41301447091812
33.38735490479619
40.39951003018536
32.58544156603253
36.881563117023504


In [9]:
MSEs = np.array(MSEs)
mean = np.mean(MSEs)
standard_deviation = np.std(MSEs)

print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

Mean: 41.65277518061606
Standard Deviation: 7.388350688563041


### While the mean and standard deviation of the dataset are 49.34 and 12.92 for a one-hidden layer NN, the said parameters are respectively 41.65 and 7.38 for a three-hidden layer NN. 

### Hence, both of the parameters decreased after adding more hidden layers.